pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go

```diff
@@ -525,21 +604,112 @@ func (lrw *logicalReplicationWriterProcessor) flushChunk(
 		batch := chunk[:min(batchSize, len(chunk))]
 		chunk = chunk[len(batch):]
 		preBatchTime := timeutil.Now()
-		s, err := bh.HandleBatch(ctx, batch)
-		if err != nil {
-			// TODO(ssd): Handle errors. We should perhaps split the batch and retry a portion of the batch.
-			// If that fails, send the failed application to the dead-letter-queue.
-			return batchStats{}, err
+
+		if s, err := bh.HandleBatch(ctx, batch); err != nil {
+			// If it already failed while applying on its own, handle the failure.
+			if len(batch) == 1 {
+				if mustProcess || !lrw.shouldPurgatory(err) {
+					if err := lrw.dlq(ctx, batch[0], err); err != nil {
+						return batchStats{}, err
+					}
+				}
+			} else {
+				// If there were multiple events in the batch, give each its own chance
+				// to apply on its own before switching to handle its failure.
+				for i := range batch {
+					if singleStats, err := bh.HandleBatch(ctx, batch[i:i+1]); err != nil {
+						if mustProcess || !lrw.shouldPurgatory(err) {
+							if err := lrw.dlq(ctx, batch[i], err); err != nil {
+								return batchStats{}, err
+							}
+						}
+					} else {
+						batch[i] = streampb.StreamEvent_KV{}
+						stats.Add(singleStats)
+					}
+				}
+			}
+		} else {
+			// Clear the event to indicate successful application.
+			for i := range batch {
+				batch[i] = streampb.StreamEvent_KV{}
+			}
+			stats.Add(s)
 		}
-		stats.Add(s)
+
 		batchTime := timeutil.Since(preBatchTime)
 		lrw.debug.RecordBatchApplied(batchTime, int64(len(batch)))
 		lrw.metrics.ApplyBatchNanosHist.RecordValue(batchTime.Nanoseconds())
 	}
 	return stats, nil
 }
 
+// shouldPurgatory returns true if a given error encountered by an attempt to
+// process an event may be resolved if processing of that event is reattempted
+// again at a later time. This could be the case, for example, if that time is
+// after the parent side of an FK relationship is ingested by another processor.
+func (lrw *logicalReplicationWriterProcessor) shouldPurgatory(err error) bool {
+	// TODO(dt): maybe this should only be constraint violation errors?
+	return true
+}
+
+const logAllDLQs = true
+
+// dlq handles a row update that fails to apply by durably recording it in a DLQ
+// or returns an error if it cannot.
+//
+// TODO(dt): implement something here.
+// TODO(dt): plumb the cdcevent.Row to this.
+func (lrw *logicalReplicationWriterProcessor) dlq(
```

Comment:
Correct me if I'm wrong, but I think we want to define the dlq as its own interface and pass in these arguments to dlq.log(). See open pr here: https://github.com/cockroachdb/cockroach/pull/126228/files

Reply:
That sounds right to me. This function is where we'll call the method on that interface to let it do its thing.

Reply:
Sounds good to me!